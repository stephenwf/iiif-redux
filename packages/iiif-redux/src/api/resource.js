import { createSelector, createStructuredSelector } from 'reselect';
import memoize from 'lodash.memoize';
import annotation from './annotation';
import canvas from './canvas';
import collection from './collection';
import externalResource from './external-resource';
import imageResource from './image-resource';
import manifest from './manifest';
import range from './range';
import resourceApi from './resource';
import sequence from './sequence';

const resourceApiMap = {
  annotation: annotation,
  canvas: canvas,
  collection: collection,
  externalResource: externalResource,
  imageResource: imageResource,
  manifest: manifest,
  range: range,
  resource: resourceApi,
  sequence: sequence,
};

const resource = memoize((selector, apiMap) => {
  return createSelector(
    selector,
    s => s,
    (entity, state) => {
      if (!entity) return null;
      const { id, schema } = entity;

      if (state.dereferenced[id] && state.dereferenced[id].loading === true) {
        return null;
      }
      if (state.dereferenced[id] && state.dereferenced[id].error) {
        return { error: state.dereferenced[id].error };
      }
      if (apiMap[schema] && resourceApiMap[schema]) {
        const api = apiMap[schema];
        const rootApi = resourceApiMap[schema];
        const selectorOrStructured = api(
          rootApi(s => {
            const resourceSchema =
              schema === 'canvas' ? 'canvases' : schema + 's';
            return s.resources[resourceSchema][id];
          })
        );

        if (typeof selectorOrStructured === 'function') {
          return selectorOrStructured(state);
        }

        const newProps = createStructuredSelector(selectorOrStructured)(state);
        newProps.schema = schema;
        return newProps;
      }
      return null;
    }
  );
});

export default resource;
