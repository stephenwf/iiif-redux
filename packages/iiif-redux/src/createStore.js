import { createStore, applyMiddleware, combineReducers, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import thunk from 'redux-thunk';
import * as reducers from './reducers';
import sagas from './sagas';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default function createCustomStore(
  customReducers = {},
  extraMiddleware = [],
  customSagas = [],
  defaultState = {}
) {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    combineReducers(reducers),
    defaultState,
    composeEnhancers(applyMiddleware(sagaMiddleware, thunk, ...extraMiddleware))
  );

  sagas.map(saga => sagaMiddleware.run(saga));
  // customSagas.map(saga => sagaMiddleware.run(saga));

  return store;
}
