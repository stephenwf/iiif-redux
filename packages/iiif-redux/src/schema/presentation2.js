/**
 * Presentation 2 - schema definition
 * ============================================================================
 *
 * - Collection
 * - Manifest
 * - Sequence
 * - Canvas
 * - Annotation
 * - AnnotationList
 * - Range
 * - Layer
 * - Image Content (Annotation)
 * - Other Content (Annotation List)
 *
 * - External resources
 * - Services
 */

// @todo Image Service API.
import { compose } from 'redux';
import { schema, normalize } from 'normalizr';
import moveStartCanvasToSequence from '../compat/moveStartCanvasToSequence';
import preprocessLinkedEntities from '../compat/preprocessLinkedEntities';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: '@id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections');
const manifest = createEntity('manifests');
const sequence = createEntity('sequences');
const canvas = createEntity('canvases');
const annotation = createEntity('annotations');
const annotationList = createEntity('annotationLists');
const range = createEntity('ranges');
const layer = createEntity('layers');
const imageResource = createEntity('imageResources'); // thumbnails + image service
const choice = createEntity('choices');

// Unofficial types?
const externalResource = createEntity('externalResources');
const service = createEntity('services', true);
const within = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    layer,
  },
  entity => {
    switch (entity['@type']) {
      case 'sc:Collection':
        return 'collection';
      case 'sc:Manifest':
        return 'manifest';
      case 'sc:Layer':
        return 'layer';
      default:
        return 'externalResource';
    }
  }
);

const rangeMember = new schema.Union(
  {
    canvas,
    range,
  },
  input => (input['@type'] === 'sc:Canvas' ? 'canvas' : 'range')
);

/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - First Sequence (Optional)
 * - Nth Sequence (Required)
 * - Canvas (Optional)
 * - Annotation (Recommended)
 * - AnnotationList (Required)
 * - Range (Optional)
 * - Layer (Optional)
 * - Image Content (Required)
 * - Other Content (Required)
 */
const RESOURCE_TYPE_MAP = {
  'sc:Collection': 'collection',
  'sc:Sequence': 'sequence',
  'sc:Manifest': 'manifest',
  'sc:Canvas': 'canvas',
  'sc:AnnotationList': 'annotationList',
  'sc:Annotation': 'annotation',
  'sc:Range': 'range',
  'sc:Layer': 'layer',
  'dctypes:Image': 'imageResource',
};
const resource = new schema.Union(
  {
    collection,
    sequence,
    manifest,
    canvas,
    annotationList,
    annotation,
    range,
    layer,
    imageResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity['@type']]) {
      return RESOURCE_TYPE_MAP[entity['@type']];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  collections: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity['@type'] === 'sc:Manifest' ? 'manifest' : 'collection')
  ),
  manifests: [manifest],
  members: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity['@type'] === 'sc:Manifest' ? 'manifest' : 'collection')
  ),

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  sequences: [sequence],
  structures: [range],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 3) Sequences
// ===========================================================================
sequence.define({
  canvases: [canvas],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,
  startCanvas: canvas,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 4) Canvases
// ===========================================================================
canvas.define({
  images: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 5) Annotation
// ===========================================================================
annotation.define({
  resource: new schema.Union(
    {
      imageResource,
      externalResource,
      choice,
    },
    input => {
      switch (input['@type']) {
        case 'dctypes:Image':
          return 'imageResource';
        case 'oa:Choice':
          return 'choice';
        default:
          return 'externalResource';
      }
    }
  ),
  on: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 6) Annotation List
// ===========================================================================
annotationList.define({
  resources: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  resources: [annotation],
  members: [rangeMember],
  canvases: [canvas],
  ranges: [range],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,
  startCanvas: canvas,
  contentLayer: layer,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 8) Layer
// ===========================================================================
layer.define({
  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,
});

// ===========================================================================
// 9) Image resource
// ===========================================================================
imageResource.define({
  service: [service],
});

// ===========================================================================
// 10) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

const preprocess = compose(
  moveStartCanvasToSequence,
  preprocessLinkedEntities,
  addMissingIds
);

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  normalizeResource as normalize,
  preprocess,
  resource,
  collection,
  manifest,
  sequence,
  canvas,
  annotation,
  annotationList,
  range,
  layer,
  imageResource,
  externalResource,
  within,
  service,
};
