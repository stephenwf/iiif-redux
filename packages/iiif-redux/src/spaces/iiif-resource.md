# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
