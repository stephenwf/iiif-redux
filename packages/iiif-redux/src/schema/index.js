import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
