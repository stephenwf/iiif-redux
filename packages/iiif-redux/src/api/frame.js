// Frame API.
import memoize from 'lodash.memoize';
import { createSelector, createStructuredSelector } from 'reselect';
import { DEFAULT_FRAME_ID } from '../spaces/frames';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';

const frame = selector => {
  // Metadata API.
  const getId = createSelector(selector, currentFrame => currentFrame.id);
  const isFocused = createSelector(
    getId,
    state => state.frames.focusedFrame,
    (currentFrameId, focusedFrameId) => focusedFrameId === currentFrameId
  );
  // @todo first extension - metadata
  // const getLabel = noop;
  // const getDescription = noop;

  // Current API.
  const getCurrentPath = createSelector(
    selector,
    currentFrame => currentFrame.path
  );
  const getCurrentPathIndex = createSelector(
    selector,
    currentFrame => currentFrame.current
  );
  const getCurrentResource = createSelector(
    getCurrentPath,
    getCurrentPathIndex,
    (currentPath, currentIndex) => currentPath[currentIndex]
  );
  const getCurrentResourceId = createSelector(
    getCurrentResource,
    currentResource => (currentResource ? currentResource.id : null)
  );
  const getCurrentResourceType = createSelector(
    getCurrentResource,
    currentResource => (currentResource ? currentResource.schema : null)
  );
  const getCurrentResourceByType = memoize(schema =>
    createSelector(getCurrentPath, path =>
      path.reduce((acc, next) => (next.schema === schema ? next : acc), null)
    )
  );

  // History API.
  const getTopLevel = createSelector(
    getCurrentPath,
    currentPath => currentPath[0] || null
  );
  const getAllTypes = createSelector(getCurrentPath, currentPath =>
    currentPath.reduce((types, path) => {
      if (types.indexOf(path.schema) === -1) {
        types.push(path.schema);
      }
      return types;
    }, [])
  );
  const canGoBack = createSelector(
    getCurrentPathIndex,
    pathIndex => pathIndex > 0
  );
  const canGoForward = createSelector(
    getCurrentPath,
    getCurrentPathIndex,
    (currentPath, currentIndex) =>
      currentPath.length === 0 ? false : currentPath.length - 1 > currentIndex
  );
  const getHistory = createSelector(
    getCurrentPath,
    getCurrentPathIndex,
    (currentPath, currentIndex) =>
      currentPath.length ? currentPath.slice(0, currentIndex) : []
  );
  const getFullHistory = getCurrentPath;
  const canGoBackToType = memoize(schema =>
    createSelector(
      getHistory,
      history =>
        history.filter(pathItem => pathItem.schema === schema).length > 0
    )
  );

  // Extensions API.
  const getAllExtensionsInternal = createSelector(
    selector,
    currentFrame => currentFrame.extensions
  );
  const getAllExtensionIds = createSelector(selector, currentFrame =>
    Object.keys(currentFrame.extensions)
  );
  const mapSingleExtension = (state, allExtensions, id) => {
    if (!allExtensions[id]) {
      return { id };
    }
    if (
      typeof allExtensions[id] === 'string' &&
      state[id] &&
      state[id][allExtensions[id]]
    ) {
      return { id, config: state[id][allExtensions[id]] };
    }
    return { id, config: allExtensions[id] };
  };
  const getAllExtensions = createSelector(
    getAllExtensionIds,
    getAllExtensionsInternal,
    state => state,
    (extensionIds, allExtensions, state) => {
      return extensionIds.map(id =>
        mapSingleExtension(state, allExtensions, id)
      );
    }
  );
  const getExtensionById = id =>
    createSelector(
      getAllExtensionsInternal,
      state => state,
      (allExtensions, state) => mapSingleExtension(state, allExtensions, id)
    );

  const getEnabledExtensionIds = createSelector(
    selector,
    currentFrame => currentFrame.enabledExtensions
  );
  const isExtensionEnabled = memoize(id =>
    createSelector(
      getEnabledExtensionIds,
      enabledExtensions => enabledExtensions.indexOf(id) !== -1
    )
  );
  const getDisabledExtensionIds = createSelector(
    getAllExtensionIds,
    getEnabledExtensionIds,
    (allExtensionIds, enabledExtensionIds) =>
      allExtensionIds.reduce((disabled, ext) => {
        if (enabledExtensionIds.indexOf(ext) === -1) {
          disabled.push(ext);
        }
        return disabled;
      }, [])
  );
  const getDisabledExtensions = createSelector(
    getDisabledExtensionIds,
    getAllExtensionsInternal,
    state => state,
    (extensionIds, allExtensions, state) => {
      return extensionIds.map(id =>
        mapSingleExtension(state, allExtensions, id)
      );
    }
  );
  const getEnabledExtensions = createSelector(
    getEnabledExtensionIds,
    getAllExtensionsInternal,
    state => state,
    (extensionIds, allExtensions, state) => {
      return extensionIds.map(id =>
        mapSingleExtension(state, allExtensions, id)
      );
    }
  );
  // const getExtensionApi = noop; // @todo come back to.

  // @todo Some resource helpers.
  // const isNavigatingCanvases = noop;
  // const getCurrentCanvasId = noop;
  // const getCurrentCanvas = noop;
  // const isNavigatingManifests = noop;
  // const getCurrentManifestId = noop;
  // const getCurrentManifest = noop;

  return {
    // Metadata API.
    getId,
    isFocused,

    // Current API.
    getCurrentPath,
    getCurrentPathIndex,
    getCurrentResource,
    getCurrentResourceId,
    getCurrentResourceType,
    getCurrentResourceByType,

    // History API.
    getTopLevel,
    getAllTypes,
    canGoBack,
    canGoForward,
    getHistory,
    getFullHistory,
    canGoBackToType,

    // Extensions API.
    getAllExtensionIds,
    getAllExtensions,
    getExtensionById,
    isExtensionEnabled,
    getDisabledExtensions,
    getEnabledExtensions,
    // getExtensionApi,

    // Some resource helpers.
    // isNavigatingCanvases,
    // getCurrentCanvasId,
    // getCurrentCanvas,
    // isNavigatingManifests,
    // getCurrentManifestId,
    // getCurrentManifest,
  };
};

export default frame;

// Allows for Window interface.
export const frameByIdSelector = memoize(
  (callable, { getId = null } = {}) => (state, props) => {
    const id = getId
      ? getId(state, props)
      : props
        ? props.frameId
        : DEFAULT_FRAME_ID;

    const selectorOrStructure = callable(frame(s => s.frames.list[id]));

    return (selectorOrStructure &&
    {}.toString.call(selectorOrStructure) === '[object Function]'
      ? selector => passThroughState => selector(passThroughState)
      : createStructuredSelector)(selectorOrStructure)(state);
  }
);

export const getFocusedFrameId = state => state.frames.focusedFrame;

export const getAllFrames = state => state.frames.list;

export const getAllFrameIds = createSelector(getAllFrames, frames =>
  Object.keys(frames)
);

export const getFocusedFrame = createSelector(
  getFocusedFrameId,
  getAllFrames,
  (id, frames) => frames[id]
);

export const focusedFrame = api => (state, props) => {
  const focusedFrameId = getFocusedFrameId(state);
  if (!focusedFrameId) {
    return null;
  }
  return createStructuredSelector(api(frame(getFocusedFrame)))(state, props);
};

export const frames = resourceListSelectorFactory(getAllFrames, frame);

/*
@todo check feasibility of this. Shouldn't be needed with simplified model.

export function registerFrameExtension(
  name,
  getFrameState = null,
  provideCustomApi = null,
  actions = {}
) {
  // name is the ID used for the KEY. Could be URI?
  // getFrameState is the `(extensionId, frameId) => state => state.myCustomRedux[extensionId]`
  // provide custom API is a IIIF Redux standard selector format. `api => selector => state => stateSubset`
  // actions will be wrapped in dispatch and given to component, possibly?
}

export function connectFramebehaviors(
  connect,
  extensions,
  mapStateToProps = null,
  mapDispatchToProps = null
) {
  // HOC for injecting actions + state, itself will wrap connect.
}
*/
