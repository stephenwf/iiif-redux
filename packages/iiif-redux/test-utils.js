import createStore from './src/createStore';
import {
  selectCanvas,
  selectManifest,
  selectSequence,
} from './src/spaces/routing';
import * as currentManifest from './src/api/current-manifest';
import * as currentSequence from './src/api/current-sequence';
import * as currentCanvas from './src/api/current-canvas';
import * as reducers from './src/reducers';
import { combineReducers } from 'redux';

export function waitForRequest(store, id) {
  return new Promise(resolve => {
    store.subscribe(() => {
      const newState = store.getState();
      if (
        newState.dereferenced &&
        newState.dereferenced[id] &&
        newState.dereferenced[id].loading === false
      ) {
        resolve();
      }
    });
  });
}

const reducer = combineReducers(reducers);
export function makeStateFromActions(...actions) {
  return [{ type: '@@redux/INIT' }, ...actions].reduce(reducer, {});
}

export async function createStoreAndImportManifest(manifestJson) {
  const id = manifestJson['@id'];
  const store = createStore();
  fetch.mockResponseOnce(JSON.stringify(manifestJson));

  const whenRequestFinishes = waitForRequest(store, id);

  store.dispatch(selectManifest({ id }));

  await whenRequestFinishes;

  const sequences = currentManifest.getSequenceIds(store.getState());
  store.dispatch(selectSequence({ id: sequences[0] }));

  const canvases = currentSequence.getCanvasIds(store.getState());
  store.dispatch(selectCanvas({ id: canvases[0] }));

  return store;
}
