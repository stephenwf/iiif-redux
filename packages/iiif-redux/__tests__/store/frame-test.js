import createStore from '../../src/createStore';
import nlsTop from '../fixtures/nls-top';
import bridges from '../fixtures/bridges';
import { waitForRequest } from '../../test-utils';
import { iiifResourceRequest } from '../../src/spaces/iiif-resource';
import {
  frameCreate,
  frameGoBack,
  frameGoToResource,
} from '../../src/spaces/frames';
import { frameByIdSelector } from '../../src/api/frame';
import { collection, manifest } from '../../src/schema/presentation2';
import resource from '../../src/api/resource';

describe('iiif/store frames', () => {
  global.fetch = require('jest-fetch-mock');

  async function loadCollection(store, collectionJson) {
    fetch.mockResponseOnce(JSON.stringify(collectionJson));
    const whenRequestFinishes = waitForRequest(store, collectionJson['@id']);
    store.dispatch(
      iiifResourceRequest(
        collectionJson['@id'],
        ['COLLECTION_REQUEST', 'COLLECTION_SUCCESS', 'COLLECTION_ERROR'],
        collection
      )
    );
    await whenRequestFinishes;
  }

  test('collections can be returned from selectors in frames', async () => {
    const store = createStore();

    store.dispatch(
      frameCreate({
        resourceId: nlsTop['@id'],
        resourceType: 'collection',
      })
    );
    // Load collection
    const finishedLoading = loadCollection(store, nlsTop);
    // @todo assertion?
    await finishedLoading;

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {
          collection: currentCollection => ({
            id: currentCollection.getId,
            type: currentCollection.getType,
            label: currentCollection.getLabel,
          }),
        }),
      }))(store.getState())
    ).toEqual({
      resource: {
        id: 'https://view.nls.uk/collections/top.json',
        label: [
          {
            '@language': 'en',
            '@value': 'National Library of Scotland IIIF Collections',
          },
        ],
        schema: 'collection',
        type: 'sc:Collection',
      },
    });

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {
          collection: currentCollection => currentCollection.getId,
        }),
      }))(store.getState())
    ).toEqual({ resource: 'https://view.nls.uk/collections/top.json' });

    expect(
      frameByIdSelector(
        api => ({
          resource: resource(api.getCurrentResource, {
            collection: currentCollection => currentCollection.getId,
          }),
        }),
        {
          getId: () => 'default-frame',
        }
      )(store.getState())
    ).toEqual({ resource: 'https://view.nls.uk/collections/top.json' });

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {
          collection: currentCollection => currentCollection.getId,
        }),
      }))(store.getState(), { frameId: 'default-frame' })
    ).toEqual({ resource: 'https://view.nls.uk/collections/top.json' });

    expect(
      frameByIdSelector(api => api.getCurrentResource)(store.getState())
    ).toEqual({
      id: 'https://view.nls.uk/collections/top.json',
      schema: 'collection',
    });
  });

  test('collections can be returned from selectors in frames', async () => {
    const store = createStore();

    store.dispatch(
      frameCreate({
        resourceId: nlsTop['@id'],
        resourceType: 'collection',
      })
    );
    // Load collection
    const finishedLoading = loadCollection(store, nlsTop);
    // @todo assertion?
    await finishedLoading;

    store.dispatch(
      frameGoToResource({
        resourceId: 'https://view.nls.uk/manifest/7446/74464614/manifest.json',
        resourceType: 'manifest',
      })
    );

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {
          collection: currentCollection => ({
            label: currentCollection.getLabel,
          }),
          manifest: currentManifest => ({ label: currentManifest.getLabel }),
        }),
      }))(store.getState())
    ).toEqual({
      resource: {
        label: [
          {
            '@language': 'en',
            '@value': 'Poems, chiefly in the Scottish dialect',
          },
        ],
        schema: 'manifest',
      },
    });
  });
  test('collections can be returned from selectors in frames', async () => {
    const store = createStore();

    store.dispatch(
      frameCreate({
        resourceId: bridges['@id'],
        resourceType: 'manifest',
      })
    );

    fetch.mockResponseOnce(JSON.stringify(bridges));
    const whenRequestFinishes = waitForRequest(store, bridges['@id']);
    store.dispatch(
      iiifResourceRequest(
        bridges['@id'],
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );
    await whenRequestFinishes;

    store.dispatch(
      frameGoToResource({
        resourceId: 'https://view.nls.uk/iiif/7446/74464117/canvas/1',
        resourceType: 'canvas',
      })
    );

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {
          manifest: currentManifest => ({ label: currentManifest.getLabel }),
          canvas: currentCanvas => ({ label: currentCanvas.getLabel }),
        }),
      }))(store.getState())
    ).toEqual({
      resource: {
        label: [{ '@language': 'en', '@value': '1' }],
        schema: 'canvas',
      },
    });

    store.dispatch(frameGoBack());

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {
          manifest: currentManifest => ({ label: currentManifest.getLabel }),
          canvas: currentCanvas => ({ label: currentCanvas.getLabel }),
        }),
      }))(store.getState())
    ).toEqual({
      resource: {
        label: [
          {
            '@language': 'en',
            '@value': 'Forth Bridge illustrations 1886-1887',
          },
        ],
        schema: 'manifest',
      },
    });

    expect(
      frameByIdSelector(api => ({
        resource: resource(api.getCurrentResource, {}),
      }))(store.getState())
    ).toEqual({ resource: null });
  });
});
