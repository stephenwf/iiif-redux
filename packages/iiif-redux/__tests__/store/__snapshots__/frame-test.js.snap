// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iiif/store frames frame resources without types will automatically be searched for and added: after 1`] = `
Object {
  "focusedFrame": "default-frame",
  "list": Object {
    "default-frame": Object {
      "config": Object {},
      "current": 0,
      "enabledExtensions": Array [],
      "extensions": Object {},
      "id": "default-frame",
      "path": Array [
        Object {
          "id": "https://view.nls.uk/manifest/7446/74464117/manifest.json",
          "schema": "manifest",
        },
      ],
    },
  },
}
`;

exports[`iiif/store frames frame resources without types will automatically be searched for and added: before 1`] = `
Object {
  "focusedFrame": "default-frame",
  "list": Object {
    "default-frame": Object {
      "config": Object {},
      "current": 0,
      "enabledExtensions": Array [],
      "extensions": Object {},
      "id": "default-frame",
      "path": Array [
        Object {
          "id": "https://view.nls.uk/manifest/7446/74464117/manifest.json",
          "schema": undefined,
        },
      ],
    },
  },
}
`;
